### PR TITLE
tools/ninja: update to 1.12.1

### PR DESCRIPTION
**Pull request by neheb**
Switch to new jobserver PR. The old one does not apply anymore.

ping @mcprat

### CODE REVIEW

**Comment by mcprat** [comment]
let's not bump the version and swap patches in the same commit...

I have a branch with everything split up into different commits but it uses a draft before I posted the PR, and with minor edits after. I can rework that if you want.

Also, in case they actually accept this version, let's wait a little longer

**Comment by mcprat** [comment]
see my PR for ninja which is an improved rework of Martin's...

also, for a cleaner diff, let's introduce the new patch with a different filename (in my test branches I just made it shorter)

**Comment by neheb** [comment]
That patch doesn't apply to the release version.

**Comment by mcprat** [comment]
at this point none of them do...

**Comment by mcprat** [comment]
oh, you mean it needs to be refreshed manually, yeah... they are very small adjustments though because of line wrapping differences and alphabetical order of headers

**Comment by Neustradamus** [comment]
To follow this PR :)

**Comment by mcprat** [comment]
so recently the 6th PR for jobserver support just opened.... now they're almost as many PRs as there were years waiting for this to be official 

even if my version won't be the final accepted one I'm glad to have stimulated some progress...

I'm still of the opinion that we should bring in whatever gets merged into ninja main, even if it needs a manual refresh, instead of going incrementally through the various implementations.

**Comment by robimarko** [comment]
Considering that the current Ninja does not work with Python 3.13, I would like to get the bump merged.
https://github.com/openwrt/openwrt/issues/16675

Are there any blockers here?

**Comment by neheb** [comment]
Not really. The patchset here works fine. Nothing has been merged to ninja yet and probably won’t be for at least a year given their glacial pace of development.

**Comment by hauke** [comment]
does the jobserver patch also works with older make versions? It says it is for make 4.4 which was released October 2022.

**Comment by dhewg** [comment]
It would be nice to mention the used jobserver PR, as there're at least 3...

The latest iteration looks to be:
https://github.com/ninja-build/ninja/pull/2506

Which was created after this PR one was, so probably worth using instead?

File: tools/ninja/patches/100-make_jobserver_support.patch [hunk 1]
```diff
@@ -363,104 +294,67 @@ Fixes https://github.com/ninja-build/ninja/issues/1139
...
++void Jobserver::Init() {
++  assert(fd_ < 0);
 +
-+bool GNUmakeTokenPool::Setup(bool ignore,
-+                             bool verbose,
-+                             double& max_load_average) {
-+  const char* value = GetEnv("MAKEFLAGS");
-+  if (!value)
-+    return false;
++  if (!ParseJobserverAuth("fifo")) {
```
Comment (mcprat) [inline tools/ninja/patches/100-make_jobserver_support.patch:310]: Martin's implementation indeed only supports Make 4.4

Comment (hauke) [reply tools/ninja/patches/100-make_jobserver_support.patch:310]: If the other versions of this patch also support older make versions I would prefer that. Even Ubuntu 24.10 is still using make 4.3 same for Debian sid.

**Comment by dhewg** [comment]
On 12/10/2024 3:51 pm, Hauke Mehrtens wrote:
> ***@***.**** commented on this pull request.
> If the other versions of this patch also support older make versions I 
> would prefer that. Even Ubuntu 24.10 is still using make 4.3 same for 
> Debian sid.

Looks like it does, from the PRs src/jobserver.h:

/// kModeDefault is the default mode to enable on the current platform.
///    This is an alias for kModeWin32Semaphore on Windows ,and
///    kModePipe on Posix, since FIFO mode was only introduced in
///    GNU Make 4.4, and older clients may not support it.

**Comment by neheb** [comment]
OK, swapped the patchset to the latest one that has the highest chance of being merged since it's a google employee.

**Comment by mcprat** [comment]
> OK, swapped the patchset to the latest one that has the highest chance of being merged since it's a google employee.

probably the right idea, however I would still like to see update as separate commit from switching patches, as well as a new filename for the new patch, otherwise this is a huge mess and the diff by itself is unreadable...

see #16693 for an update while keeping the same patch (light rework was required of course)

after that, rebasing this should be as easy as make a copy, ignore conflicts, rebase, paste back, refresh,
then we can focus on just the patch itself

**Comment by mcprat** [comment]
you didnt have to close this, did you not want to work on testing the new jobserver patch?

**Comment by neheb** [comment]
nah.

**Pull request closed without merging**